File: app/src/main/java/com/pigdogbay/foodhygieneratings/model/Injector.java
package com.pigdogbay.foodhygieneratings.model;

import android.content.Context;

import com.pigdogbay.foodhygieneratings.R;
import com.pigdogbay.lib.utils.ActivityUtils;

import org.json.JSONObject;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class Injector {

    private boolean isBuilt = false;
    private static Injector injector;
    private MainModel mainModel;
    private List<LocalAuthority> localAuthorities;

    public static MainModel getMainModel() {
        return injector.mainModel;
    }

    public static Injector getInjector() {
        if (injector == null){
            injector = new Injector();
        }
        return injector;
    }

    public void build(Context context){
        if (isBuilt){
            return;
        }
        isBuilt = true;

        Context applicationContext = context.getApplicationContext();

        mainModel = new MainModel();
        mainModel.setDataProvider(dummyDataProvider(applicationContext));
        //mainModel.setDataProvider(new WebDataProvider());
    }

    private MainModel.IDataProvider dummyDataProvider(Context context){
        try {
            String data = ActivityUtils.readResource(context, R.raw.stoke);
            return new DummyDataProvider(data);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public List<LocalAuthority> getLocalAuthorities(Context context){
        if (localAuthorities==null) {
            try {
                String data = ActivityUtils.readResource(context.getApplicationContext(), R.raw.authorities);
                JSONObject jsonObject = new JSONObject(data);
                localAuthorities = FoodHygieneAPI.parseAuthorities(jsonObject);
            } catch (Exception e) {
                e.printStackTrace();
                localAuthorities =  new ArrayList<>();
            }
        }
        return localAuthorities;
    }



}

File: app/src/main/java/com/pigdogbay/foodhygieneratings/model/Injector.kt
package com.pigdogbay.foodhygieneratings.model

import android.content.Context

import com.pigdogbay.foodhygieneratings.R
import com.pigdogbay.lib.utils.ActivityUtils
import com.pigdogbay.lib.utils.BitmapUtils
import com.pigdogbay.lib.utils.PreferencesHelper

import org.json.JSONObject

import java.io.IOException
import java.util.ArrayList


object Injector {

    private var isBuilt = false
    lateinit var mainModel: MainModel
    lateinit var settings: Settings
    private var localAuthorities: List<LocalAuthority>? = null

    fun build(context: Context) {
        if (isBuilt) {
            return
        }
        isBuilt = true

        val applicationContext = context.applicationContext
        val preferencesHelper = PreferencesHelper(applicationContext)
        settings = Settings(preferencesHelper)

        mainModel = MainModel()
        mainModel.dataProvider = WebDataProvider()
    }

    fun createFetcher(context: Context) : IPlaceFetcher {
        return DummyPlace(BitmapUtils.getBitmap(context, R.drawable.fhis_pass_and_eat_safe))
    }

    private fun dummyDataProvider(context: Context): MainModel.IDataProvider {
        try {
            val data = ActivityUtils.readResource(context, R.raw.stoke)
            return DummyDataProvider(data)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }

    }

    /*
        To update R.raw.authorities, go to
        http://api.ratings.food.gov.uk/Help/Api/GET-Authorities-basic
        Press test API and copy JSON into https://jsonformatter.org/json-pretty-print

        To get a more detailed authorities info see
        http://ratings.food.gov.uk/authorities/json

     */
    fun getLocalAuthorities(context: Context): List<LocalAuthority> {
        if (localAuthorities == null) {
            try {
                val data = ActivityUtils.readResource(context.applicationContext, R.raw.authorities)
                val jsonObject = JSONObject(data)
                localAuthorities = FoodHygieneAPI.parseAuthorities(jsonObject)
            } catch (e: Exception) {
                e.printStackTrace()
                localAuthorities = ArrayList()
            }

        }
        return localAuthorities!!
    }
}
